bisum/tests/test_bisum.py

```diff
@@ -6,98 +6,70 @@
-
-### acc -> ac
-### cdd -> cd  ----> accd ----> acd
-
-##### if slicing returns an empty array avoid lexsort!!!!!!!!!!!!!!!! AND "dad,mom -> dm"
-einsumstr = "aaa,wop -> a" ## SPARSE-SPARSE, YES post-inter-externals, YES post-transpose
-A = torch.rand(13,13,13, device=device)
-B = torch.rand(10,10,10, device=device)
-print( torch.allclose( bisum( einsumstr, A.to_sparse(), B.to_sparse()).to_dense(), torch.einsum( einsumstr , A, B )) )
-
-
-##### if slicing returns an empty array avoid lexsort!!!!!!!!!!!!!!!! AND "dad,mom -> dm"
-einsumstr = "aaa,wop -> a" ## SPARSE-DENSE, YES post-inter-externals, YES post-transpose
-A = torch.rand(10,10,10, device=device)
-B = torch.rand(10,10,10, device=device)
-print( torch.allclose( bisum( einsumstr, A.to_sparse(), B).to_dense(), torch.einsum( einsumstr , A, B )) )
-
-##### if slicing returns an empty array avoid lexsort!!!!!!!!!!!!!!!! AND "dad,mom -> dm"
-einsumstr = "qaq,wow -> " ## SPARSE-DENSE, YES post-inter-externals, YES post-transpose
-A = torch.rand(10,10,10, device=device)
-B = torch.rand(10,10,10, device=device)
-print( torch.allclose( bisum( einsumstr, A.to_sparse(), B).to_dense(), torch.einsum( einsumstr , A, B )) )
\ No newline at end of file
+
+
+# set up tensors for all tests (do NOT modify the tensors), treat them
+# as readonly; if tensors should be changed then change scope to
+# "function"
+@pytest.fixture(scope="module")
+def A_big():
+    return torch.rand(10,10,10,10,10,10, device=device)
+
+@pytest.fixture(scope="module")
+def A():
+    return torch.rand(10,10,10, device=device)
+
+@pytest.fixture(scope="module")
+def B():
+    return torch.rand(10,10,10, device=device)
+
+@pytest.fixture(scope="module")
+def C():
+    return torch.rand(13,13,13, device=device)
+
+
+
+@pytest.mark.parametrize("einsumstr", [
+    "aeacec,cdd -> aed",  #SPARSE-SPARSE, NO post-inter-externals, NO post-transpose
+])
+def test_sparse_sparse_no_post_inter_externals(A_big, B, einsumstr):
+    assert torch.allclose(
+        bisum(einsumstr, A_big.to_sparse(), B.to_sparse()).to_dense(),
+        torch.einsum(einsumstr, A_big, B))
+
+@pytest.mark.parametrize("einsumstr", [
+    "daa,aed -> a", # SPARSE-SPARSE, YES post-inter-externals, NO post-transpose
+    "daa,aed -> ea", ## SPARSE-SPARSE, YES post-inter-externals, YES post-transpose
+])
+def test_sparse_sparse_with_post_inter_externals(A, B, einsumstr):
+    ### acc -> ac
+    ### cdd -> cd  ----> accd ----> acd
+
+    ### acc -> ac
+    ### cdd -> cd  ----> accd ----> acd
+
+    assert torch.allclose(
+        bisum(einsumstr, A.to_sparse(), B.to_sparse()).to_dense(),
+        torch.einsum(einsumstr, A, B))
+
+
+def test_sparse_sparse_with_empty_array_1(C, B):
+    # if slicing returns an empty array avoid lexsort!!!!!!!!!!!!!!!! AND "dad,mom -> dm"
+    einsumstr = "aaa,wop -> a" ## SPARSE-SPARSE, YES post-inter-externals, YES post-transpose
```

Comment:
You can also parametrize this tests if you like. I just kept it as a single test because I didn't know what was special about this one compared to the `test_sparse_dense_with_empty_array_2()` one.